pkg/provisioner/resourcehandlers.go

```diff
@@ -131,6 +131,53 @@ func createObjectBucket(ob *v1alpha1.ObjectBucket, c versioned.Interface, retryI
 	return
 }
 
+func updateObjectBucket(ob *v1alpha1.ObjectBucket, c versioned.Interface, retryInterval, retryTimeout time.Duration) (result *v1alpha1.ObjectBucket, err error) {
+	logD.Info("updating ObjectBucket", "name", ob.Name)
+	result, err = c.ObjectbucketV1alpha1().ObjectBuckets().Update(context.TODO(), ob, metav1.UpdateOptions{})
+	if err != nil {
+		return nil, err
+	}
+	return result, err
+}
+
+func getObjectBucketFromClaimKey(key string, obc *v1alpha1.ObjectBucketClaim, c versioned.Interface, retryInterval, retryTimeout time.Duration) (ob *v1alpha1.ObjectBucket, err error) {
+	obName, err := objectBucketNameFromClaimKey(key)
+	if err != nil {
+		return nil, err
+	}
+
+	logD.Info("seeing if object bucket for claim exists", "expected object bucket name", obName)
+	ob, err = c.ObjectbucketV1alpha1().ObjectBuckets().Get(context.TODO(), obName, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Not found is a valid exit here which returns no error and a nil result
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	emptyRef := corev1.ObjectReference{}
+	if ob.Spec.ClaimRef == nil || *ob.Spec.ClaimRef == emptyRef {
+		// If the CRD validation for the OB improperly specifies the object reference field, it will
+		// be missing. Do our best to recover from this case gracefully.
+		logD.Info("found OB matching this OBC that does not have an object reference, assuming it should belong to this claim", "OB", ob.Name)
+		ob.Spec.ClaimRef, err = claimRefForKey(key, c)
+		if err != nil {
+			return nil, fmt.Errorf("error generating OB %q's reference to OBC: %v", ob.Name, err)
+		}
+		ob, err = updateObjectBucket(ob, c, defaultRetryBaseInterval, defaultRetryTimeout)
+		if err != nil {
+			return nil, fmt.Errorf("error updating OB %q with assumed claim ref: %v", ob.Name, err)
+		}
+	} else if bucketIsOwnedByClaim(obc, ob) {
+		logD.Info("found OB belonging to this OBC", "OB", ob.Name)
+	} else {
+		return nil, fmt.Errorf("found OB %q not owned by OBC %q, likely an artifact from a previous OBC that did not get cleaned up, user must delete the OB in order to allow OBC provisioning to continue", ob.Name, key)
+	}
+
+	return ob, err
+}
```

Comment:
This is a nit but I'd restructure as:
```
...
   if bucketIsOwnedByClaim(obc, ob) {
         logD.Info(sprintf("found OB %q matching OBC %s/%s but the OB's ClaimRef is empty. Binding OB to OBC", ob.Name, obc.Namespace, obc.Name))
         return ob, err
   }

   if ob.Spec.ClaimRef == nil || *ob.Spec.ClaimRef == emptyRef {
		// If the CRD validation for the OB improperly specifies the object reference field, it will
		// be missing. Do our best to recover from this case gracefully.
		logD.Info("found OB matching this OBC that does not have an object reference, assuming it should belong to this claim", "OB", ob.Name)
		ob.Spec.ClaimRef, err = claimRefForKey(key, c)
		if err != nil {
		      err = fmt.Sprintf("error generating OB %q's reference to OBC: %v", ob.Name, err)
		} else
		       ob, err = updateObjectBucket(ob, c, defaultRetryBaseInterval, defaultRetryTimeout)
		             if err != nil {
			            err = fmt.Sprintf("error updating OB %q with assumed claim ref: %v", ob.Name, err)
		             }
   } else {
          err = fmt.Sprintf("found OB %q not owned by OBC %q, likely an artifact from a previous OBC that did not get cleaned up, user must delete the OB in order to allow OBC provisioning to continue: %v", ob.Name, key, err) 
   }
   return nil, err
}
```

Reply:
The restructure you suggested mixed up the logic a bit, but I restructured in the spirit of your comment.

Reply:
In thinking about this more last night, I'm not sure I like it anymore. I was hoping to reduce the number or returns but I don't want to mess up the logic. Feel free to ignore!